src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs

```diff
@@ -479,22 +479,22 @@ private static async Task<object> ExecuteHubMethod(ObjectMethodExecutor methodEx
         private void InitializeHub(THub hub, HubConnectionContext connection)
         {
             hub.Clients = new HubCallerClients(_hubContext.Clients, connection.ConnectionId);
-            hub.Context = new DefaultHubCallerContext(connection);
+            hub.Context = connection.HubCallerContext;
             hub.Groups = _hubContext.Groups;
         }
 
-        private Task<bool> IsHubMethodAuthorized(IServiceProvider provider, ClaimsPrincipal principal, IList<IAuthorizeData> policies)
+        private Task<bool> IsHubMethodAuthorized(IServiceProvider provider, HubConnectionContext hubConnectionContext, IList<IAuthorizeData> policies, string hubMethodName, object[] hubMethodArguments)
         {
             // If there are no policies we don't need to run auth
             if (!policies.Any())
             {
                 return TaskCache.True;
             }
 
-            return IsHubMethodAuthorizedSlow(provider, principal, policies);
+            return IsHubMethodAuthorizedSlow(provider, hubConnectionContext.User, policies, new AuthHubConnectionContext(hubConnectionContext, hubMethodName, hubMethodArguments));
```

Comment:
Noo, allocations...

Reply:
Haha, I know, it's draft for a reason

Reply:
The tricky thing is figuring out how to avoid this allocation. We could pool them maybe but that's a bunch of work. What about skipping this whole path if there are no policies?

Reply:
>What about skipping this whole path if there are no policies?

We already do that https://github.com/aspnet/AspNetCore/pull/11070#discussion-diff-292252496R489

Reply:
Does that temper your concern about allocations @davidfowl ? If there is no auth policy on a hub method we won't allocate.

Reply:
Made it no longer allocate, but it does rely on serial hub invocations

Reply:
Honestly I prefer the allocation but I'm not @davidfowl ;). We already know we don't want to have serial hub invocations long-term, so adding more things that depend on that behavior is just setting us up to regress perf later.

Reply:
I'd rather do the "correct" thing in terms of our design plans and then **pro**gress perf later by finding a way to reduce the allocations.

Reply:
Better to allocate.